File: frontend/src/stores/project.ts
/**
 * B2 Showcase Pinia store(冻结骨架,subagent 不允许改 state/getter 名,只能加 action)。
 *
 * 数据源:GET /api/projects/{name}/{manifest,tree,bom,assembly,pipeline}
 * 切项目时清缓存重拉。
 */
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import axios from 'axios'

// ── 类型契约(对应 backend/schemas/project.py) ───────────────────────────────

export interface AssemblyPart {
  id: string
  name: string
  glb: string
  step: string
  transform: { translation: number[]; rotation: number[] }
  explode_offset: number[]
  color: string
  owner: string
  cad_only?: boolean
  missing?: boolean
}

export interface AssemblyGroup {
  id: string
  name: string
  parts: string[]
}

export interface Deliverable {
  kind: string
  path: string
  owner: string
  extra?: Record<string, unknown>
}

export interface ManifestSummary {
  mass_g: number
  dof: number
  parts_count: number
  cost_by_category: Record<string, number>
  currency: string
  bbox: number[]
}

export interface AssemblyManifest {
  project: string
  name: string
  version: string
  updated_at: string | null
  tags: string[]
  hero_image: string
  summary: ManifestSummary
  assembly: { parts: AssemblyPart[]; groups: AssemblyGroup[] }
  deliverables: Deliverable[]
  fallback: boolean
}

export interface FileNode {
  path: string
  kind: string
  size: number
  mtime: string | null
  children?: FileNode[]
}

export interface BomVendor {
  name: string
  url: string
  price_cny: number
  tier: 'pro' | 'maker' | 'budget' | 'instant'
}

export interface BomItem {
  category: string
  subcategory: string
  name: string
  qty: number
  unit_price: number
  total: number
  datasheet?: string
  vendors: BomVendor[]
  selected_vendor: string
}

export interface BomDoc {
  currency: string
  items: BomItem[]
  summary: { total: number; by_category: Record<string, number> }
}

export interface AssemblyStep {
  id: string
  text: string
  parts: number
  refs: string[]
}

export interface AssemblyPhase {
  name: string
  icon: string
  steps: AssemblyStep[]
}

export interface AssemblyDoc {
  tools: string[]
  assumptions: string[]
  phases: AssemblyPhase[]
}

export interface PipelineSnapshot {
  task_id: string
  nodes: any[]
  edges: any[]
}

// ── Connectivity (B2-connectivity-view.md §2.1) ────────────────────────────

export type NodeKind =
  | 'mcu' | 'sensor' | 'actuator' | 'power' | 'module' | 'display'
  | 'cad_part' | 'actuator_cross_domain' | 'generic'

export type EdgeKind = 'mechanical' | 'power' | 'data'
export type InterfaceKind = 'data' | 'power' | 'mechanical'

export interface NodeInterface {
  id: string
  kind: InterfaceKind
}

export interface ConnectivityRef {
  bom?: string | null
  datasheet?: string | null
  schematic_block?: string | null
  step?: string | null
  glb?: string | null
  part_meta?: string | null
  cad_model?: string | null
}

export interface ConnectivityNode {
  id: string
  kind: NodeKind
  label: string
  domain: string
  owner: string
  owner_label: string
  ref: ConnectivityRef
  interfaces: NodeInterface[]
}

export interface ConnectivityEdge {
  id: string
  from: string             // "<node_id>:<interface_id>"
  to: string
  kind: EdgeKind
  label: string
  data_subtype: string
}

export interface ConnectivityDoc {
  version: string
  generated_at: string | null
  nodes: ConnectivityNode[]
  edges: ConnectivityEdge[]
  merge_failed: boolean
}

// ── store 定义 ─────────────────────────────────────────────────────────────

const API_BASE = '/api/projects'

export const useProjectStore = defineStore('project', () => {
  // state — 名字冻结,subagent 不改
  const current = ref<string>('')
  const manifest = ref<AssemblyManifest | null>(null)
  const tree = ref<FileNode[]>([])
  const bom = ref<BomDoc | null>(null)
  const assemblyDoc = ref<AssemblyDoc | null>(null)
  const pipeline = ref<PipelineSnapshot | null>(null)
  const connectivity = ref<ConnectivityDoc | null>(null)  // B2-connectivity-view
  const loading = ref(false)
  const error = ref<string | null>(null)

  // B2-connectivity-view §3.6 — 抽屉点击「跳转 §4 装配视图」时,
  // 由 ShowcaseHomeView 监听这个 ref,在 3D viewer 里高亮对应 part。
  // 仅 connectivity 流程使用,加 action 不改既有 state 名。
  const connectivityHighlight = ref<string | null>(null)
  function setConnectivityHighlight(partId: string | null) {
    connectivityHighlight.value = partId
  }

  // getters
  const partsById = computed(() => {
    const m: Record<string, AssemblyPart> = {}
    manifest.value?.assembly.parts.forEach(p => { m[p.id] = p })
    return m
  })
  const deliverableByKind = computed(() => {
    const m: Record<string, Deliverable> = {}
    manifest.value?.deliverables.forEach(d => { m[d.kind] = d })
    return m
  })
  const totalCost = computed(() => bom.value?.summary?.total ?? 0)
  const isFallback = computed(() => manifest.value?.fallback === true)

  // actions
  async function ensureLoaded(project: string) {
    if (current.value === project && manifest.value) return
    loading.value = true
    error.value = null
    current.value = project
    try {
      const [m, t, b, p] = await Promise.allSettled([
        axios.get<AssemblyManifest>(`${API_BASE}/${project}/manifest`),
        axios.get<FileNode[]>(`${API_BASE}/${project}/tree`),
        axios.get<BomDoc>(`${API_BASE}/${project}/bom`).catch(() => null),
        axios.get<PipelineSnapshot>(`${API_BASE}/${project}/pipeline`).catch(() => null),
      ])
      manifest.value = m.status === 'fulfilled' ? m.value.data : null
      tree.value = t.status === 'fulfilled' ? t.value.data : []
      bom.value = b && b.status === 'fulfilled' && b.value ? b.value.data : null
      pipeline.value = p && p.status === 'fulfilled' && p.value ? p.value.data : null

      // assembly.json 单独拉(可能 404)
      try {
        const a = await axios.get<AssemblyDoc>(`${API_BASE}/${project}/assembly`)
        assemblyDoc.value = a.data
      } catch {
        assemblyDoc.value = null
      }

      // connectivity.json (B2-connectivity-view) — 端点总返回 200(空 doc 兜底)
      try {
        const c = await axios.get<ConnectivityDoc>(`${API_BASE}/${project}/connectivity`)
        connectivity.value = c.data
      } catch {
        connectivity.value = null
      }

      if (m.status === 'rejected') {
        error.value = '无法加载 manifest'
      }
    } catch (exc: any) {
      error.value = exc?.message ?? '未知错误'
    } finally {
      loading.value = false
    }
  }

  function clear() {
    current.value = ''
    manifest.value = null
    tree.value = []
    bom.value = null
    assemblyDoc.value = null
    pipeline.value = null
    connectivity.value = null
    error.value = null
  }

  function fileUrl(path: string): string {
    if (!current.value) return ''
    return `${API_BASE}/${current.value}/file?path=${encodeURIComponent(path)}`
  }

  // ── B2.7: SSE 实时刷新 ───────────────────────────────────────────────────
  // 订阅 /api/events,收到属于当前项目的 task 事件后,debounced 重拉 manifest/tree。
  let _sse: EventSource | null = null
  let _refreshTimer: number | null = null

  function _scheduleRefresh() {
    if (_refreshTimer !== null) clearTimeout(_refreshTimer)
    _refreshTimer = window.setTimeout(async () => {
      const proj = current.value
      if (!proj) return
      try {
        const [m, t] = await Promise.all([
          axios.get<AssemblyManifest>(`${API_BASE}/${proj}/manifest`),
          axios.get<FileNode[]>(`${API_BASE}/${proj}/tree`),
        ])
        manifest.value = m.data
        tree.value = t.data
      } catch {
        // 忽略,下个事件再试
      }
    }, 800) as unknown as number
  }

  function startLiveRefresh() {
    if (_sse) return
    if (typeof window === 'undefined' || typeof EventSource === 'undefined') return
    try {
      _sse = new EventSource('/api/events')
      _sse.onmessage = (ev) => {
        // 简单策略:任意 task_events 都 debounced 重拉(轻量,不阻塞 UI)
        if (!current.value) return
        try {
          const data = JSON.parse(ev.data || '{}')
          if (typeof data === 'object') _scheduleRefresh()
        } catch {
          _scheduleRefresh()
        }
      }
      _sse.onerror = () => {
        // 连接抖动 — EventSource 自带重连,不主动 close
      }
    } catch {
      _sse = null
    }
  }

  function stopLiveRefresh() {
    if (_sse) {
      _sse.close()
      _sse = null
    }
    if (_refreshTimer !== null) {
      clearTimeout(_refreshTimer)
      _refreshTimer = null
    }
  }

  return {
    // state
    current, manifest, tree, bom, assemblyDoc, pipeline, connectivity, loading, error,
    connectivityHighlight,
    // getters
    partsById, deliverableByKind, totalCost, isFallback,
    // actions
    ensureLoaded, clear, fileUrl,
    startLiveRefresh, stopLiveRefresh,
    setConnectivityHighlight,
  }
})

File: backend/schemas/project.py
"""B2 项目展示页数据契约(schema freeze)。

对应 doc/design/B2-showcase-frontend.md §2.2 / §2.3 / §2.5。

冻结点(并行 subagent 必须遵守):
- ManifestRead(含 assembly.parts[] / deliverables[] / summary / tags)
- BomDoc(items[].category 12 类强枚举 / vendors[] >=2)
- AssemblyDoc(phases[] 5 段 + tools / assumptions)
- TreeNode(path/kind/size/mtime)
- PipelineSnapshot(nodes/edges,前端 elkjs 算坐标)
"""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


# ── BOM ────────────────────────────────────────────────────────────────────────

BomCategory = Literal[
    "microcontroller", "sensor", "actuator", "power", "module", "display",
    "structural", "enclosure", "mechanism", "hardware", "3D-printed", "generic",
]
VendorTier = Literal["pro", "maker", "budget", "instant"]


class BomVendor(BaseModel):
    name: str
    url: str = ""
    price_cny: float
    tier: VendorTier = "pro"


class BomItem(BaseModel):
    category: BomCategory
    subcategory: str = ""
    name: str
    qty: int
    unit_price: float
    total: float
    datasheet: str = ""
    vendors: list[BomVendor] = Field(default_factory=list)
    selected_vendor: str = ""


class BomSummary(BaseModel):
    total: float
    by_category: dict[str, float] = Field(default_factory=dict)


class BomDoc(BaseModel):
    currency: str = "CNY"
    items: list[BomItem] = Field(default_factory=list)
    summary: BomSummary = Field(default_factory=lambda: BomSummary(total=0.0))


# ── Manifest ───────────────────────────────────────────────────────────────────

class AssemblyPart(BaseModel):
    id: str
    name: str
    glb: str = ""
    step: str = ""
    transform: dict = Field(default_factory=lambda: {
        "translation": [0, 0, 0],
        "rotation": [0, 0, 0, 1],
    })
    explode_offset: list[float] = Field(default_factory=lambda: [0.0, 0.0, 0.0])
    color: str = "#a0a0a0"
    owner: str = ""
    cad_only: bool = False
    missing: bool = False


class AssemblyGroup(BaseModel):
    id: str
    name: str
    parts: list[str] = Field(default_factory=list)


class Assembly(BaseModel):
    parts: list[AssemblyPart] = Field(default_factory=list)
    groups: list[AssemblyGroup] = Field(default_factory=list)


DeliverableKind = Literal[
    "prd", "cad", "schematic", "pcb", "firmware", "algorithm",
    "bom", "test_report", "video", "status", "urdf", "sim_video", "image",
]


class Deliverable(BaseModel):
    kind: DeliverableKind
    path: str
    owner: str = ""
    extra: dict = Field(default_factory=dict)


class ManifestSummary(BaseModel):
    mass_g: int = 0
    dof: int = 0
    parts_count: int = 0
    cost_by_category: dict[str, float] = Field(default_factory=dict)
    currency: str = "CNY"
    bbox: list[float] = Field(default_factory=lambda: [0.0, 0.0, 0.0])


class ManifestRead(BaseModel):
    project: str
    name: str = ""
    version: str = "0.0.1"
    updated_at: datetime | None = None

    tags: list[str] = Field(default_factory=list)
    hero_image: str = ""
    summary: ManifestSummary = Field(default_factory=ManifestSummary)

    assembly: Assembly = Field(default_factory=Assembly)
    deliverables: list[Deliverable] = Field(default_factory=list)

    fallback: bool = False  # True = 后端目录扫描兜底,非 PM 确认


# ── Tree ──────────────────────────────────────────────────────────────────────

FileKind = Literal[
    "cad", "model3d", "schematic_src", "pcb_src", "kicad_project",
    "gerber", "archive", "netlist", "csv",
    "markdown", "pdf", "code", "json", "yaml",
    "image",
    "dir", "binary",
]


class TreeNode(BaseModel):
    path: str
    kind: FileKind
    size: int = 0
    mtime: datetime | None = None
    children: list["TreeNode"] = Field(default_factory=list)


# ── Assembly instructions ─────────────────────────────────────────────────────

class AssemblyStep(BaseModel):
    id: str
    text: str
    parts: int = 0
    refs: list[str] = Field(default_factory=list)


class AssemblyPhase(BaseModel):
    name: str
    icon: str = ""
    steps: list[AssemblyStep] = Field(default_factory=list)


class AssemblyDoc(BaseModel):
    tools: list[str] = Field(default_factory=list)
    assumptions: list[str] = Field(default_factory=list)
    phases: list[AssemblyPhase] = Field(default_factory=list)


# ── Pipeline (Workflow 视图) ──────────────────────────────────────────────────

class WorkflowPort(BaseModel):
    id: str
    label: str
    type: Literal["task", "doc", "cad", "schematic", "pcb", "code", "data", "signal"]
    position: Literal["left", "right"] = "left"


class WorkflowField(BaseModel):
    label: str
    value: str
    variant: Literal["select", "readonly", "badge"] = "readonly"


class WorkflowNodeData(BaseModel):
    seq: int
    title: str
    subtitle: str = ""
    headerColor: str = "#6366f1"
    owner: str
    status: Literal["pending", "running", "done", "failed"] = "pending"
    duration_ms: int | None = None
    inputs: list[WorkflowPort] = Field(default_factory=list)
    outputs: list[WorkflowPort] = Field(default_factory=list)
    fields: list[WorkflowField] = Field(default_factory=list)
    deliverable: dict = Field(default_factory=dict)


class WorkflowNode(BaseModel):
    id: str
    type: str = "deliverable"
    position: dict = Field(default_factory=lambda: {"x": 0, "y": 0})
    data: WorkflowNodeData


class WorkflowEdge(BaseModel):
    id: str
    source: str
    sourceHandle: str
    target: str
    targetHandle: str
    type: str = "typed"
    data: dict = Field(default_factory=dict)


class PipelineSnapshot(BaseModel):
    task_id: str = ""
    nodes: list[WorkflowNode] = Field(default_factory=list)
    edges: list[WorkflowEdge] = Field(default_factory=list)


# ── Connectivity (B2-connectivity-view §2.1) ─────────────────────────────────

NodeKind = Literal[
    "mcu", "sensor", "actuator", "power", "module", "display",
    "cad_part", "actuator_cross_domain", "generic",
    # B2-connectivity-view 扩展(2026-05-20):算法 / PCB 板 / 固件
    "algorithm", "pcb", "firmware",
]
EdgeKind = Literal["mechanical", "power", "data"]
InterfaceKind = Literal["data", "power", "mechanical"]


class NodeInterface(BaseModel):
    id: str
    kind: InterfaceKind


class ConnectivityRef(BaseModel):
    bom: str | None = None
    datasheet: str | None = None
    schematic_block: str | None = None
    step: str | None = None
    glb: str | None = None
    part_meta: str | None = None
    cad_model: str | None = None


class ConnectivityNode(BaseModel):
    id: str
    kind: NodeKind
    label: str
    domain: str = ""
    owner: str
    owner_label: str = ""
    ref: ConnectivityRef = Field(default_factory=ConnectivityRef)
    interfaces: list[NodeInterface] = Field(default_factory=list)


class ConnectivityEdge(BaseModel):
    id: str
    from_: str = Field(alias="from")
    to: str
    kind: EdgeKind
    label: str = ""
    data_subtype: str = ""

    model_config = {"populate_by_name": True}


class ConnectivityDoc(BaseModel):
    version: str = "1.0"
    generated_at: datetime | None = None
    nodes: list[ConnectivityNode] = Field(default_factory=list)
    edges: list[ConnectivityEdge] = Field(default_factory=list)
    merge_failed: bool = False


# ── 项目列表条目 ────────────────────────────────────────────────────────────────

class ProjectListItem(BaseModel):
    name: str
    title: str = ""
    has_manifest: bool = False
    has_charter: bool = False
    parts_count: int = 0
    updated_at: datetime | None = None
